crates/ruff_python_parser/src/token_source.rs

```diff
@@ -108,29 +117,43 @@ impl<'src> TokenSource<'src> {
         TokenSourceCheckpoint {
             lexer: self.lexer.checkpoint(),
             tokens_position: self.tokens.len(),
+            comments_position: self.comments.len(),
         }
     }
 
     /// Restore the token source to the given checkpoint.
     pub(crate) fn rewind(&mut self, checkpoint: TokenSourceCheckpoint<'src>) {
         self.lexer.rewind(checkpoint.lexer);
         self.tokens.truncate(checkpoint.tokens_position);
+        self.comments.truncate(checkpoint.comments_position);
     }
 
     /// Consumes the token source, returning the collected tokens and any errors encountered during
     /// lexing. The token collection includes both the trivia and non-trivia tokens.
-    pub(crate) fn finish(self) -> (Vec<Token>, Vec<LexicalError>) {
+    pub(crate) fn finish(self) -> (Vec<Token>, CommentRanges, Vec<LexicalError>) {
         assert_eq!(
             self.current_kind(),
             TokenKind::EndOfFile,
             "TokenSource was not fully consumed"
         );
 
-        (self.tokens, self.lexer.finish())
+        let comment_ranges = CommentRanges::new(self.comments);
+        (self.tokens, comment_ranges, self.lexer.finish())
     }
 }
 
 pub(crate) struct TokenSourceCheckpoint<'src> {
     lexer: LexerCheckpoint<'src>,
```

Comment:
What's the sizse of a `TokenSourceCheckpoint` and `ParserCheckpoint` at this point? It seems to becoming somewhat big ;)

Reply:
Yeah, it is getting quite big:
* `ParserCheckpoint` - 176 bytes
* `TokenSourceCheckpoint` - 152 bytes
* `LexerCheckpoint` - 136 bytes